### PR TITLE
Improve `ElementSubdomainModifier` moving boundary and reinitialization

### PR DESCRIPTION
**Pull request by Wendy-Ji**
This pull request addresses several problems with `ElementSubdomainModifier`. The biggest changes in this PR are:

- The creation of the `ElementSubdomainModifierBase` to contain all the code for changing subdomain, modifying sidesets/nodesets, and reinitialization (`ElementSubdomainModifier` now only determines which elements are changing subdomain, and their previous and new subdomain IDs)
- Greater control over the changing sidesets/nodesets using the `moving_boundaries` and `moving_boundary_subdomain_pairs` parameters (`moving_boundary_name` and `complement_moving_boundary_name` are replaced by these)
-  Control over which changed elements are reinitialized using the `reinitialize_subdomains` and `old_subdomain_reinitialized` parameters (`apply_initial_conditions` is replaced by these)

This fixes:
- The `moving_boundary_name` was not updating correctly and the `complement_moving_boundary_name` was being used instead. This redundancy has been removed, and the ability to choose which 'side' of the moving boundary to use is still retained
- An adaptivity file (`*-s0*.e`) file was being output at every time step even where there were no subdomain changes. `meshChanged()` is now only called when at least one element changes its subdomain. 
- Segmentation fault for particular scenarios when running in parallel has been fixed

This adds:
- `moving_boundary_subdomain_pairs` gives greater control over the updates of the boundary. This is based on which pairs of subdomains elements are between (only applies to elements that change subdomain ID). External element sides can now also be added to the boundary.
- Reinitialization of variables and auxvariables was previously all or nothing for elements that changed subdomains. Now user can control whether reinitialization occurs depending on the old and new subdomain IDs
- Reinitialization of material properties always occurred previously. Now they reinitialize on the same elements as the variables and auxvariables do

refs https://github.com/idaholab/moose/discussions/25736

@hugary1995 Continues some of the work from https://github.com/idaholab/moose/pull/26254

### CODE REVIEW

**Comment by Wendy-Ji** [comment]
Hello, this PR will cause some errors when testing `malamute` and `racoon` as the parameters have been changed for `CoupledVarThresholdElementSubdomainModifier`. I have made PRs to patch these. Thank you.

https://github.com/idaholab/malamute/pull/160
https://github.com/hugary1995/raccoon/pull/171

File: modules/stochastic_tools/examples/paper/results/weak.tex
Comment (hugary1995) [inline modules/stochastic_tools/examples/paper/results/weak.tex:15]: not sure what's changed here... is github broken?

File: python/MooseDocs/test/gold/latex/extensions/bibtex.tex
Comment (hugary1995) [inline python/MooseDocs/test/gold/latex/extensions/bibtex.tex:8]: come on github...

File: python/MooseDocs/test/gold/latex/extensions/config.tex
Comment (hugary1995) [inline python/MooseDocs/test/gold/latex/extensions/config.tex:3]: Also let's restore this change.

File: python/MooseDocs/test/gold/latex/extensions/core.tex
Comment (hugary1995) [inline python/MooseDocs/test/gold/latex/extensions/core.tex:1]: Also restore this file.

File: python/MooseDocs/test/gold/latex/extensions/folder/index.tex
Comment (hugary1995) [inline python/MooseDocs/test/gold/latex/extensions/folder/index.tex:1]: Restore this file.

File: python/MooseDocs/test/gold/latex/extensions/include.tex
Comment (hugary1995) [inline python/MooseDocs/test/gold/latex/extensions/include.tex:1]: Restore.

File: python/MooseDocs/test/gold/latex/extensions/materialicon.tex
Comment (hugary1995) [inline python/MooseDocs/test/gold/latex/extensions/materialicon.tex:1]: Restore.

File: python/MooseDocs/test/gold/latex/extensions/plotly.tex
Comment (hugary1995) [inline python/MooseDocs/test/gold/latex/extensions/plotly.tex:1]: Restore.

File: python/MooseDocs/test/gold/latex/extensions/style.tex
Comment (hugary1995) [inline python/MooseDocs/test/gold/latex/extensions/style.tex:1]: Restore.

**Review by hugary1995 — CHANGES_REQUESTED** [review]
I have some really minor comments. There are some files that need to be restored -- I will take care of them.

**Comment by hugary1995** [comment]
Since I am not an independent reviewer as I authored part of `ElementSubdomainModifierBase.h`/`.C`, we will need at least one approval from either @dewenyushu or @GiudGiud .

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (dewenyushu) [inline framework/src/meshmodifiers/ElementSubdomainModifierBase.C:62]: It is a little counter intuitive here, I guess the idea is to prevent reinitialize elements multiple times if it is in the `reinitialize_subdomains`? Then maybe call this as something like "allow_duplicate_reinit"?

File: framework/include/meshmodifiers/ElementSubdomainModifier.h
Comment (Wendy-Ji) [reply framework/include/meshmodifiers/ElementSubdomainModifier.h:36]: Haha, I'm pretty sure this is yours. It is a nice docstring

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (dewenyushu) [inline framework/src/meshmodifiers/ElementSubdomainModifierBase.C:157]: Do we now support moving to a new subdomain? - this would be a great news for additive manufacturing simulations :)

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (dewenyushu) [inline framework/src/meshmodifiers/ElementSubdomainModifierBase.C:413]: Two comments here: 
1. This is not just removing inactive moving boundary. Looks like this functions also checks active sides in the elements' family and update the boundary accordingly. So a different name would be better?
2. If a boundary does not contain any sides in one step, the boundary ID still exist in the mesh. So in later steps new sides can be added to this boundary if needed, right? Do you happen to know how this case is handled during output and visualization ( in Exodus and Paraview)?

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (dewenyushu) [inline framework/src/meshmodifiers/ElementSubdomainModifierBase.C:510]: Will it be helpful to add an option to not update old and older solutions here?

**Review by dewenyushu — CHANGES_REQUESTED** [review]
Excited to see this code improvement. Very well coded and documented. I just have a few comments.

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (Wendy-Ji) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:62]: Hello, sorry, this parameter name is a bit confusing and I'm open to suggestions for it. `previous_subdomain_reinitialized` is an optional parameter that gives greater control over which elements are reinitialized. In addition to checking whether an element's new subdomain is in the `reinitialize_subdomains`, the element's old subdomain will also be checked. If `previous_subdomain_reinitialized` is set to false, then that element will only be reinitialized if its old subdomain isn't in `reinitialize_subdomains` (see test file `reinitialization_from_into`). This is useful, for example, when you want to reinitialize elements that move from inactive to active subdomains, but ignore any elements that change from one active subdomain to another active subdomain.

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (Wendy-Ji) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:413]: 1. A different name would be better. This is specifically for updating the boundary for adaptive mesh refinement, as the boundaries need to be updated from the parent to the children elements. This will be changed to `updateAMRMovingBoundary` to better reflect its purpose. If you have a better idea, that would also be good.
2. That is a good point. I will add a test case to check that new sides can still be added to the boundary.

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (Wendy-Ji) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:510]: Currently, if an old solution does not exist, then the `setOldAndOlderSolutions` isn't carried out. 

If the user specifies that reinitialization occurs in an element that changes subdomain, it was assumed that the old solution of the previous subdomain wasn't required anymore, as the element would be a 'new' material, e.g. solid to liquid. 

There is still the option to turn off reinitialization completely, if the user only wants to change subdomain but wants to keep all the values the same. 

Is there a scenario you're thinking of where it would be useful to reinitialize an element value, but still keep the old and older solutions as the values from the previous subdomain? If there is, then that option could be added in.

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (Wendy-Ji) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:157]: I think moving into a completely new subdomain was introduced in the `add_subdomain_ids` and `add_subdomain_names` in a different PR. And moving from one subdomain to another subdomain was introduced in the `TimedSubdomainModifier`. My additive manufacturing models are also much happier with these additions : )

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (dewenyushu) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:413]: I like `updateAMRMovingBoundary` :) and thanks for agreeing to check that special case

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (dewenyushu) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:510]: I was thinking about phase transition too. But you are right, it makes more sense to reinit both old and older solutions here. I don't have a strong application case in mind at the moment, so it might be best to leave it as is!

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (dewenyushu) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:62]: Thanks for your detailed explanation! Now I get your point. I think it may be okay to keep the name as you have, but maybe add a bit more explanation about the default behavior when `previous_subdomain_reinitialized =true`?

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (Wendy-Ji) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C]: Hello, I have added a test case `return_moving_boundary`. The boundary is defined between subdomains 1 and 2. Subdomain 1 changes to 3 then back to 1, which means the boundary does not contain any sides in a timestep, but does return at the later timestep. This looks fine in the Exodus and Paraview. Let me know if you were thinking of a different test.

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (Wendy-Ji) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C]: I have updated the docstring for the parameter to hopefully be clearer (I also changed the parameter's name to `old_subdomain_reinitialized`, mainly to make it shorter)

File: libmesh
Comment (Wendy-Ji) [reply libmesh]: Oops. I changed the libmesh back

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (dewenyushu) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C]: Thanks! Looks good to me!

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (dewenyushu) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C]: Yes, that is exactly what I meant. Thanks for adding the test!

**Comment by GiudGiud** [comment]
with regards to that last commit deleting the images, it s not enough to add a commit to delete the images. they are sitll in the commit history of the PR. you need to git rebase then edit the commit adding the images to truly remove them and the memory they take

File: large_media
Comment (Wendy-Ji) [reply large_media:1]: large_media PR for images used in ESM markdowns

**Review by hugary1995 — APPROVED** [review]
Looks great to me. Thanks!

**Comment by hugary1995** [comment]
> with regards to that last commit deleting the images, it s not enough to add a commit to delete the images. they are sitll in the commit history of the PR. you need to git rebase then edit the commit adding the images to truly remove them and the memory they take

I believe this is resolved since all commits are squashed into one. Let me know otherwise.

**Comment by GiudGiud** [comment]
I can do the independent CCB review tonight or tomorrow

**Comment by hugary1995** [comment]
> I can do the independent CCB review tonight or tomorrow

That'd be great. Thanks!

**Comment by GiudGiud** [comment]
Will be tomorrow, got too busy

File: framework/doc/content/source/meshmodifiers/ElementSubdomainModifier.md
Comment (GiudGiud) [inline framework/doc/content/source/meshmodifiers/ElementSubdomainModifier.md:5]:
```suggestion
The `ElementSubdomainModifier` modifies an element subdomain ID. This class is inherited by other mesh modifiers, such as [CoupledVarThresholdElementSubdomainModifier.md] and [TimedSubdomainModifier.md], which implement different criteria for a subdomain ID to be modified.
```

File: framework/doc/content/source/meshmodifiers/ElementSubdomainModifier.md
Comment (GiudGiud) [inline framework/doc/content/source/meshmodifiers/ElementSubdomainModifier.md:28]: this ensures the param names stay in sync with the docs. no need to do it for every single instance of each parameter, but each parameter is a good idea
```suggestion
The change of element subdomains will alter the definitions of certain sidesets and nodesets. The parameters [!param](/MeshModifiers/CoupledVarThresholdElementSubdomainModifier/moving_boundaries) and `moving_boundary_subdomain_pairs` can optionally be used to modify the corresponding sidesets/nodesets over the elements that change subdomain. The names of the boundaries are specified in `moving_boundaries`, and the pair of subdomains that each boundary in `moving_boundaries` lies between must be specified in the corresponding `moving_boundary_subdomain_pairs`. The element side from the first subdomain of the pair is added to the boundary.
```

File: framework/doc/content/source/meshmodifiers/TimedSubdomainModifier.md
Comment (GiudGiud) [inline framework/doc/content/source/meshmodifiers/TimedSubdomainModifier.md:12]: off topic, but I m curious if the nodeset modification works with TimedSubdomainMod
there isnt as much of a front iirc, an entire subdomain can get swapped

File: framework/include/meshmodifiers/ElementSubdomainModifierBase.h
Comment (GiudGiud) [inline framework/include/meshmodifiers/ElementSubdomainModifierBase.h:16]:
```suggestion
/**
 * Base class for mesh modifiers modifying element subdomains
 */
class ElementSubdomainModifierBase : public ElementUserObject
```

File: framework/include/meshmodifiers/ElementSubdomainModifierBase.h
Comment (GiudGiud) [inline framework/include/meshmodifiers/ElementSubdomainModifierBase.h:94]: add docstring (comment)

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (GiudGiud) [inline framework/src/meshmodifiers/ElementSubdomainModifierBase.C:32]:
```suggestion
  params.addDeprecatedParam<BoundaryName>(
```

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (GiudGiud) [inline framework/src/meshmodifiers/ElementSubdomainModifierBase.C:35]:
```suggestion
  params.addDeprecatedParam<BoundaryName>(
```

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (GiudGiud) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:32]: will remove it from auto-suggest in vscode plugin

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (GiudGiud) [inline framework/src/meshmodifiers/ElementSubdomainModifierBase.C:189]: can we move these 6 lines above? if we did not change the elements we didnt move the sideset/nodeset?

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (GiudGiud) [inline framework/src/meshmodifiers/ElementSubdomainModifierBase.C:535]: let's add a mooseAssert that we only have a single NL system. even though the UO is associated with a system, I bet we would prefer all systems do it

**Review by GiudGiud — COMMENTED** [review]
looks good

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (GiudGiud) [inline framework/src/meshmodifiers/ElementSubdomainModifierBase.C:543]: it's a good thought.
We should add an assert here, querying the problem for whether it just failed a time step (if we have such an API)

File: framework/doc/content/source/meshmodifiers/TimedSubdomainModifier.md
Comment (Wendy-Ji) [reply framework/doc/content/source/meshmodifiers/TimedSubdomainModifier.md:12]: There is an example in the tests called 'return_moving_boundary', which has a boundary defined between subdomains 1 and 2. It uses the `TimedSubdomainModifier` to change subdomain 1 to 3, and then back, and the nodeset reflects that change.

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (Wendy-Ji) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:32]: Hi. I didn't use the `DeprecatedParam` here as these parameters (`moving_boundary_name` and `complement_moving_boundary_name`) cannot be used anymore. Instead, using these will lead to an error message explaining to the user the updated parameters. I did use `DeprecatedParam` for the `apply_initial_conditions`, since that one can still be used.

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (Wendy-Ji) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:35]: Same as above comment

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (Wendy-Ji) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C]: Hello, I have swapped the order

**Review by dewenyushu — APPROVED** [review]

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (GiudGiud) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:544]: actually, seems like it would be problematic to put an error there. Maybe just a "flagSolutionInvalid" would be enough.
Anyway this is out of scope here

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (GiudGiud) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:32]: It does not hurt to use deprecatedParam here.
and the nice thing is that it removes it from the documentation

File: framework/doc/content/source/meshmodifiers/TimedSubdomainModifier.md
Comment (GiudGiud) [reply framework/doc/content/source/meshmodifiers/TimedSubdomainModifier.md:12]: great!

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (GiudGiud) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C:537]: added

**Review by GiudGiud — APPROVED** [review]

File: framework/src/meshmodifiers/ElementSubdomainModifierBase.C
Comment (GiudGiud) [reply framework/src/meshmodifiers/ElementSubdomainModifierBase.C]: added

**Comment by GiudGiud** [comment]
@Wendy-Ji if you could update your malamute patch please. If you don't have time please just give me write access and I'll finish it up

**Comment by Wendy-Ji** [comment]
> @Wendy-Ji if you could update your malamute patch please. If you don't have time please just give me write access and I'll finish it up

The PR is still under review, but I think everything in the patch is up to date now. I've given you write access in case there's something else I've missed

**Pull request merged**
